src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -113,41 +120,34 @@ private void SetEntry(CacheEntry entry)
             // Initialize the last access timestamp at the time the entry is added
             entry.LastAccessed = utcNow;
 
-            var added = false;
             CacheEntry priorEntry;
-
-            _entryLock.EnterWriteLock();
-            try
+            if (_entries.TryRemove(entry.Key, out priorEntry))
```

Comment:
I think the most important thing is to add a test which verifies that the current `ConcurrentDictionary` behavior is incorrect, and the new behavior is correct.  Since it's a race condition, you can't have a unit test which guarantees failure, but I think one thread reading while another is writing should hit the race quickly enought.